README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# belg

<!-- badges: start -->

[![CRAN
status](http://www.r-pkg.org/badges/version/belg)](https://cran.r-project.org/package=belg)
[![R build
status](https://github.com/r-spatialecology/belg/workflows/pkgdown-and-test-coverage/badge.svg)](https://github.com/r-spatialecology/belg/actions)
[![codecov](https://app.codecov.io/gh/r-spatialecology/belg/branch/master/graph/badge.svg)](https://app.codecov.io/gh/r-spatialecology/belg)
[![CRAN RStudio mirror
downloads](http://cranlogs.r-pkg.org/badges/belg)](https://cran.r-project.org/package=belg)
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1209419.svg)](https://doi.org/10.5281/zenodo.1209419)
<!-- badges: end -->

Boltzmann entropy (also called configurational entropy) has been
recently adopted to analyze entropy of landscape gradients (Gao et
al. (2017, 2018, 2019)). The goal of **belg** is to provide an efficient
C++ implementation of this method in R. It also extend the original idea
by allowing calculations on data with missing values (Nowosad and Gao
(2020)).

## Installation

You can install the released version of belg from
[CRAN](https://CRAN.R-project.org) with:

``` r
install.packages("belg")
```

And the development version from [GitHub](https://github.com/) with:

``` r
# install.packages("remotes")
remotes::install_github("r-spatialecology/belg")
```

## Examples

As an example, we use two rasters - `land_gradient1` representing a
complex landscape and `land_gradient2` representing a simple landscape:

``` r
library(raster)
library(belg)
land_gradient1 = raster(system.file("raster/land_gradient1.tif", package = "belg"))
land_gradient2 = raster(system.file("raster/land_gradient2.tif", package = "belg"))
plot(stack(land_gradient1, land_gradient2))
```

<img src="man/figures/README-unnamed-chunk-1-1.png" width="100%" />

The main function in this package, `get_boltzmann()`, calculates the
Boltzmann entropy of a landscape gradient:

``` r
get_boltzmann(land_gradient1)
#> [1] 188772.5
get_boltzmann(land_gradient2)
#> [1] 121875.2
```

This function accepts a `SpatRaster`, `stars`, `RasterLayer`,
`RasterStack`, `RasterBrick`, `matrix`, or `array` object as an input.
It allows for calculation of the relative (the `relative` argument equal
to `TRUE`) and absolute Boltzmann entropy of a landscape gradient. As a
default, it uses a logarithm of base 10 (`log10`), however `log` and
`log2` are also available options for the `base` argument.

``` r
get_boltzmann(land_gradient1, base = "log")
#> [1] 434664.7
get_boltzmann(land_gradient1, relative = TRUE)
#> [1] 137645.4
get_boltzmann(land_gradient1, base = "log2", relative = TRUE)
#> [1] 457248.1
```

Two methods of calculating the Boltzmann entropy of a landscape gradient
are available: `"hierarchy"` (default) for the hierarchy-based method
(Gao et al., 2017) or `"aggregation"` for the aggregation-based method
(Gao et al., 2019). The aggregation-based method requires that the
number of rows and columns in the input data must be a multiple of 2.

``` r
get_boltzmann(land_gradient1, method = "aggregation")
#> [1] 188772.5
get_boltzmann(land_gradient1, relative = TRUE, method = "aggregation")
#> [1] 137645.4
```

More examples can be find at <https://github.com/Nowosad/belg-examples>.

## References

-   Gao, Peichao, Hong Zhang, and Zhilin Li. “A hierarchy-based solution
    to calculate the configurational entropy of landscape gradients.”
    Landscape Ecology 32(6) (2017): 1133-1146.
-   Gao, Peichao, Hong Zhang, and Zhilin Li. “An efficient analytical
    method for computing the Boltzmann entropy of a landscape gradient.”
    Transactions in GIS (2018).
-   Gao, Peichao and Zhilin Li. “Aggregation-based method for computing
    absolute Boltzmann entropy of landscape gradient with full
    thermodynamic consistency.” Landscape Ecology (2019).
-   Nowosad, J.; Gao, P. belg: A Tool for Calculating Boltzmann Entropy
    of Landscape Gradients. Entropy 2020, 22, 937.
    <https://doi.org/10.3390/e22090937>


NEWS.md
# belg 1.5.0

* Adds terra support
* Moves internal data to external files
* Cleans dependencies

# belg 1.4.0

* Adds `na_adjust` argument to `get_boltzmann()`

# belg 1.3.0

* Removes int overflow

# belg 1.2.0

* Makes aggregation method the default 

# belg 1.0.3

* Adds support for stars objects

# belg 1.0.2

* Moves raster to Imports

# belg 1.0.1

* Adds compatibility with R 4.0.0 (matrix objects now also inherit from class "array")

# belg 1.0.0

* The ["Aggregation-based method for computing absolute Boltzmann entropy of landscape gradient with full thermodynamic consistency"](https://doi.org/10.1007/s10980-019-00854-3) added
* New datasets added

# belg 0.2.1

* The first vignette added

# belg 0.2.0

* Support for data with missing values added

# belg 0.1.1

* Solaris compilation fixed

# belg 0.1.0

* NA error check added
* Negative values are shifted
* Documentation details updated

# belg 0.0.4

* Working rcpp code added

# belg 0.0.0.9000

* Added a `NEWS.md` file to track changes to the package


R/get_boltzmann.R
#' Boltzmann entropy of a landscape gradient
#'
#' Calculates the Boltzmann entropy of a landscape gradient
#'
#' @param x SpatRaster, stars, RasterLayer, RasterStack, RasterBrick, matrix, or array.
#' @param base A logarithm base ("log", "log2" or "log10").
#' @param relative Should a relative or absolute entropy be calculated? TRUE or FALSE (default).
#' @param method A method used. Either "hierarchy" for
#' the hierarchy-based method (Gao et al., 2017) or "aggregation" (default)
#' for the aggregation-based method (Gao et al., 2019).
#' @param na_adjust Should the output value be adjusted to the proportion of not missing cells? Either TRUE (default) or FALSE
#'
#' @return a numeric vector
#'
#' @details The method for computing the Boltzmann entropy of a landscape
#' gradient works on integer values that are either positive or equals to zero.
#' This function automatically rounds values to the nearest integer value
#' (rounding halfway cases away from zero) and negative values are shifted to
#' positive values.
#'
#' @references Gao, Peichao, Hong Zhang, and Zhilin Li. "A hierarchy-based
#' solution to calculate the configurational entropy of landscape gradients."
#' Landscape Ecology 32.6 (2017): 1133-1146.
#'
#' @references Gao, Peichao, Hong Zhang, and Zhilin Li. "An efficient
#' analytical method for computing the Boltzmann entropy of a landscape
#' gradient." Transactions in GIS (2018).
#'
#' @references Gao, Peichao and Zhilin Li. "Aggregation-based method
#' for computing absolute Boltzmann entropy of landscape gradient
#' with full thermodynamic consistency"
#' Landscape Ecology (2019)
#'
#' @examples
#' new_c = c(56, 86, 98, 50, 45, 56, 96, 25,
#'           15, 55, 85, 69, 12, 52, 25, 56,
#'           32, 25, 68, 98, 58, 66, 56, 58)
#'
#'
#' lg = matrix(new_c, nrow = 3, ncol = 8, byrow = TRUE)
#' get_boltzmann(lg, relative = FALSE, method = "hierarchy", base = "log10")
#' get_boltzmann(lg, relative = TRUE, method = "hierarchy", base = "log2")
#' get_boltzmann(lg, relative = TRUE, method = "hierarchy", base = "log")
#'
#' @name get_boltzmann
#' @export
get_boltzmann = function(x, method = "aggregation", na_adjust = TRUE, base = "log10", relative = FALSE) UseMethod("get_boltzmann")

#' @name get_boltzmann
#' @export
get_boltzmann.default = function(x, method = "aggregation", na_adjust = TRUE, base = "log10", relative = FALSE){
  if (method == "hierarchy"){
    result = get_boltzmann_default(x, base, relative)
  } else if (method == "aggregation"){
    result = get_boltzmann_aggregation(x, base, relative)
  }
  # if (!missing(resolution)){
  #   if (length(resolution) == 1){
  #     resolution = resolution^2
  #   } else {
  #     resolution = resolution[1] * resolution[2]
  #   }
  # } else {
  #   resolution = 1
  # }
  # if (scale == "no_of_cells"){
  #   result = (result) / (ncol(x) * nrow(x))
  # } else if (scale == "resolution"){
  #   result = (result) / (resolution) #* 1000000
  # } else if (scale == "all"){
  #   result = (result) / (ncol(x) * nrow(x) * resolution)
  # }
  if (na_adjust){
    result = (result) / (not_na_prop(x))
  }
  return(result)
}

##' @name get_boltzmann
##' @export
get_boltzmann.matrix = function(x, method = "aggregation", na_adjust = TRUE, base = "log10", relative = FALSE){
  if (method == "hierarchy"){
    result = get_boltzmann_default(x, base, relative)
  } else if (method == "aggregation"){
    result = get_boltzmann_aggregation(x, base, relative)
  }
  # if (!missing(resolution)){
  #   if (length(resolution) == 1){
  #     resolution = resolution^2
  #   } else {
  #     resolution = resolution[1] * resolution[2]
  #   }
  # } else {
  #   resolution = 1
  # }
  # if (scale == "no_of_cells"){
  #   result = (result) / (ncol(x) * nrow(x))
  # } else if (scale == "resolution"){
  #   result = (result) / (resolution) #* 1000000
  # } else if (scale == "all"){
  #   result = (result) / (ncol(x) * nrow(x) * resolution)
  # }
  if (na_adjust){
    result = (result) / (not_na_prop(x))
  }
  return(result)
}

#' @name get_boltzmann
#' @export
get_boltzmann.array = function(x, method = "aggregation", na_adjust = TRUE, base = "log10", relative = FALSE){
  if (method == "hierarchy"){
    result = apply(x, MARGIN = 3, get_boltzmann_default, base, relative)
  } else if (method == "aggregation"){
    result = apply(x, MARGIN = 3, get_boltzmann_aggregation, base, relative)
  }
  # if (!missing(resolution)){
  #   if (length(resolution) == 1){
  #     resolution = resolution^2
  #   } else {
  #     resolution = resolution[1] * resolution[2]
  #   }
  # } else {
  #   resolution = 1
  # }
  # if (scale == "no_of_cells"){
  #   result = (result) / (ncol(x) * nrow(x))
  # } else if (scale == "resolution"){
  #   result = (result) / (resolution) #* 1000000
  # } else if (scale == "all"){
  #   result = (result) / (ncol(x) * nrow(x) * resolution)
  # }
  if (na_adjust){
    result = (result) / apply(x, MARGIN = 3, not_na_prop)
  }
  return(result)
}

#' @name get_boltzmann
#' @export
get_boltzmann.RasterLayer = function(x, method = "aggregation", na_adjust = TRUE, base = "log10", relative = FALSE){
  if (!requireNamespace("sp", quietly = TRUE))
    stop("Package sp required, please install it first", call. = FALSE)
  if (!requireNamespace("raster", quietly = TRUE))
    stop("Package raster required, please install it first", call. = FALSE)
  # if (missing(resolution)){
  #   resolution = c(raster::xres(x),
  #                  raster::yres(x))
  # }
  get_boltzmann(raster::as.matrix(x), base = base, relative = relative, method = method, na_adjust = na_adjust)
}

#' @name get_boltzmann
#' @export
get_boltzmann.RasterStack = function(x, method = "aggregation", na_adjust = TRUE, base = "log10", relative = FALSE){
  if (!requireNamespace("sp", quietly = TRUE))
    stop("Package sp required, please install it first", call. = FALSE)
  if (!requireNamespace("raster", quietly = TRUE))
    stop("Package raster required, please install it first", call. = FALSE)
  # if (missing(resolution)){
  #   resolution = c(raster::xres(x),
  #                  raster::yres(x))
  # }
  get_boltzmann(raster::as.array(x), base = base, relative = relative, method = method, na_adjust = na_adjust)
}

#' @name get_boltzmann
#' @export
get_boltzmann.RasterBrick = function(x, method = "aggregation", na_adjust = FALSE, base = "log10", relative = FALSE){
  if (!requireNamespace("sp", quietly = TRUE))
    stop("Package sp required, please install it first", call. = FALSE)
  if (!requireNamespace("raster", quietly = TRUE))
    stop("Package raster required, please install it first", call. = FALSE)
  # if (missing(resolution)){
  #   resolution = c(raster::xres(x),
  #                  raster::yres(x))
  # }
  get_boltzmann(raster::as.array(x), base = base, relative = relative, method = method, na_adjust = na_adjust)
}

#' @name get_boltzmann
#' @export
get_boltzmann.stars = function(x, method = "aggregation", na_adjust = TRUE, base = "log10", relative = FALSE){
  if (!requireNamespace("stars", quietly = TRUE))
    stop("Package stars required, please install it first", call. = FALSE)
  if (length(x) > 1){
    warning("The input stars object has more than one attribute. \nBoltzmann entropy is calculated for the first attribute in the stars object only", call. = FALSE)
  }
  # if (missing(resolution)){
  #   resolution = c(abs(stars::st_dimensions(x)$x$delta),
  #                  abs(stars::st_dimensions(x)$y$delta))
  # }
  get_boltzmann(x[[1]], base = base, relative = relative, method = method, na_adjust = na_adjust)
}

#' @name get_boltzmann
#' @export
get_boltzmann.SpatRaster = function(x, method = "aggregation", na_adjust = TRUE, base = "log10", relative = FALSE){
  if (!requireNamespace("terra", quietly = TRUE))
    stop("Package terra required, please install it first", call. = FALSE)
  get_boltzmann(terra::as.array(x), base = base, relative = relative, method = method, na_adjust = na_adjust)
}

